src/parser/jobs/brd/modules/AlwaysBeCasting.ts

```diff
@@ -0,0 +1,88 @@
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import AlwaysBeCasting from 'parser/core/modules/AlwaysBeCasting'
+import {BuffEvent, CastEvent} from 'fflogs'
+
+const SONG_DURATION_MS = 30 * 1000
+
+interface ArmyWindow {
+	start: number,
+	end: number
+}
+
+export default class BardABC extends AlwaysBeCasting {
+	armyHistory: ArmyWindow[] = []
+	currentMuse: ArmyWindow | undefined
+	currentPaeon: ArmyWindow | undefined
+
+	protected init() {
+		super.init()
+		this.addEventHook('applybuff', {by: 'player', abilityId: [STATUSES.ARMYS_MUSE.id]}, this.onMuse)
+		this.addEventHook('removebuff', {by: 'player', abilityId: [STATUSES.ARMYS_MUSE.id]}, this.onRemoveMuse)
+		this.addEventHook('cast', {by: 'player', abilityId: [ACTIONS.THE_WANDERERS_MINUET.id, ACTIONS.MAGES_BALLAD.id, ACTIONS.ARMYS_PAEON.id]}, this.onSong)
+	}
+
+	private endMuse() {
+		if (this.currentMuse) {
+			this.armyHistory.push(this.currentMuse)
+			this.currentMuse = undefined
+		}
+	}
+
+	private endPaeon() {
+		if (this.currentPaeon) {
+			this.armyHistory.push(this.currentPaeon)
+			this.currentPaeon = undefined
+		}
+	}
+
+	private onMuse(event: BuffEvent) {
+		this.currentMuse = {
+			start: event.timestamp,
+			end: Math.min(this.parser.fight.end_time, event.timestamp + STATUSES.ARMYS_MUSE.duration),
+		}
+	}
+
+	private onRemoveMuse(event: BuffEvent) {
+		if (this.currentMuse) {
+			this.currentMuse.end = event.timestamp
+			this.endMuse()
+		}
+	}
+
+	private onSong(event: CastEvent) {
+		if (event.ability.guid === ACTIONS.ARMYS_PAEON.id) {
+			this.currentPaeon = {
+				start: event.timestamp,
+				end: Math.min(this.parser.fight.end_time, event.timestamp + SONG_DURATION_MS),
+			}
+		} else if (this.currentPaeon) {
+			this.currentPaeon.end = event.timestamp
+			this.endPaeon()
+		}
+	}
+
+	private isArmyBuffActive(timestamp: number): boolean {
+		return this.armyHistory.some(army => timestamp > army.start && timestamp < army.end)
+	}
+
+	protected getUptimePercent(): number {
+		const fightDuration = this.parser.currentDuration - this.downtime.getDowntime()
+		const armyDuration = this.armyHistory.reduce((acc, army) => {
+			const downtime = this.downtime.getDowntime(army.start, army.end)
+			return acc + army.end - army.start - downtime
+		}, 0)
+
+		const uptime = this.gcd.gcds.reduce((acc, gcd) => {
+			const duration = this.gcd._getGcdLength(gcd)
```

Comment:
I think this is all kinds of inaccurate but yolo

Reply:
ye